---
title: "Untitled"
output: html_document
date: "2025-10-15"
---

load dependencies
```{r Loading libraries, include=FALSE}
library(tidyverse)
library(GenomicRanges)
library(DiffBind)
library(Rsubread)
library(reshape2)
library(zoo)
library(plyr)
library(pheatmap)
library(JASPAR2020)
library(BSgenome.Ggallus.ENSEMBL.galGal6)
library(TFBSTools)
library(chromVAR)
library(tradeSeq)
library(org.Gg.eg.db)
library(DESeq2)
options(scipen = 999)
```


load data
```{r}
data = dba(sampleSheet="/data/Megan/tomoAnalysis/Epi_ATAC/EpiATAC_samples.csv")
data
```

Count reads at peaks
```{r}
peaks.gr = read.table("/data/Megan/tomoAnalysis/scATAC-seq/HH5_scATACseq_peaks.bed") %>%
  dplyr::rename("CHR"="V1", "START"="V2", "END"="V3") %>% GRanges()
peakwidths = width(peaks.gr)
peaks.gr = peaks.gr[peakwidths<5000 & peakwidths>150]
data_2 = dba.count(data,
                peaks=peaks.gr,
                summits = FALSE, score=DBA_SCORE_READS, bParallel=T)
raw_counts = dba.peakset(data_2, bRetrieve=T, DataType=DBA_DATA_FRAME)
data_3 = dba.count(data,
                peaks=peaks.gr,
                summits = FALSE, score=DBA_SCORE_NORMALIZED, bParallel=T)
data_4 = dba.normalize(data_3, normalize=DBA_NORM_LIB,
                             library=DBA_LIBSIZE_FULL)
norm_counts = dba.peakset(data_4, bRetrieve=T, DataType=DBA_DATA_FRAME)
norm_counts
```

Make peak variable name
```{r}
raw_counts = raw_counts %>%
  mutate(peak = paste0(CHR, ":", START, "-", END)) %>%
  dplyr::select(-CHR, -START, -END) %>% remove_rownames() %>%
  column_to_rownames(var = c("peak"))
```

Filter for row sums greater than 50
```{r}
keep = rowSums(raw_counts) >= 50
counts = raw_counts[keep,,]
```

build metadata table
```{r}
colData = data.frame(row.names=colnames(counts), data=colnames(counts))
colData = colData %>% separate(data, c('data', 'treatment', 'replicate'), "_")
colData
```

Make DEseq object and run PCA
```{r}
dds = DESeqDataSetFromMatrix(countData = counts,
                              colData = colData,
                              design = ~treatment)
pcaData = plotPCA(rlog(dds), intgroup=c("treatment", "replicate")) + ggplot2::theme_classic()

pcaData = plotPCA(rlog(dds), intgroup=c("treatment", "replicate"), returnData=TRUE)
ggplot(pcaData, aes(PC1, PC2, color=treatment, shape=replicate)) +
  geom_point(size=3) + coord_fixed() + theme_classic()
```


Run DESeq
```{r}
dds = DESeq(dds)
DMSO_CHIR_resDf = as.data.frame(results(dds, contrast=c("treatment","CHIR","DMSO")))
DMSO_XAV_resDf = as.data.frame(results(dds, contrast=c("treatment","XAV","DMSO")))
DMSO_CHIR_resDf$peak = rownames(DMSO_CHIR_resDf)
DMSO_XAV_resDf$peak = rownames(DMSO_XAV_resDf)

resDf = inner_join(DMSO_CHIR_resDf, DMSO_XAV_resDf, by="peak")
resDf
```


Load anterior and posterior peaksets
```{r}
anterior_peaks = read.table("anterior_peaks_TradeSeq.bed")
anterior_peaks = anterior_peaks %>% unite("peak", V1:V2, remove=TRUE, sep=":") %>% unite("peak", peak:V3, remove=TRUE, sep="-")
posterior_peaks = read.table("posterior_peaks_TradeSeq.bed")
posterior_peaks = posterior_peaks %>% unite("peak", V1:V2, remove=TRUE, sep=":") %>% unite("peak", peak:V3, remove=TRUE, sep="-")
```


Get WNT-responsive peaks
```{r}
upCHIR_downXAV = dplyr::filter(resDf, log2FoldChange.x > 0.5 & log2FoldChange.y < -0.5)
downCHIR_upXAV = dplyr::filter(resDf, log2FoldChange.x < -0.5 & log2FoldChange.y > 0.5)
```


Plot scatter plot of WNT-responsive peaks
```{r}
upCHIR_downXAV_plot = ggplot() + theme_classic() + 
  geom_hline(yintercept=0, linetype="dashed", size=0.5) +
  geom_vline(xintercept=0, linetype="dashed", size=0.5) +
  labs(x="DMSO vs CHIR", y="DMSO vs XAV") +
  geom_point(data=resDf, aes(x=log2FoldChange.x, y=log2FoldChange.y), alpha=0.2, color="gray") +
  geom_point(data=dplyr::filter(resDf, peak %in% upCHIR_downXAV$peak), 
             aes(x=log2FoldChange.x, y=log2FoldChange.y), alpha=0.2, color="#9D0041")
upCHIR_downXAV_plot

downCHIR_upXAV_plot = ggplot() + theme_classic() + 
  geom_hline(yintercept=0, linetype="dashed", size=0.5) +
  geom_vline(xintercept=0, linetype="dashed", size=0.5) +
  labs(x="DMSO vs CHIR", y="DMSO vs XAV") +
  geom_point(data=resDf, aes(x=log2FoldChange.x, y=log2FoldChange.y), alpha=0.2, color="gray") +
  geom_point(data=dplyr::filter(resDf, peak %in% downCHIR_upXAV$peak), 
             aes(x=log2FoldChange.x, y=log2FoldChange.y), alpha=0.2, color="#054D80")
downCHIR_upXAV_plot
```

load lef/bcat-bound peaks
```{r}
bCat_Lef1_peaks = read.table("bCat_Lef1_HH5_peaks_new.bed")
bCat_Lef1_GR = bCat_Lef1_peaks %>% dplyr::rename("CHR"="V1", "START"="V2", "END"="V3") %>% GRanges()

resDf_GR = resDf %>% separate(peak, c("CHR", "START", "END")) %>% 
  dplyr::select(CHR, START, END) %>% GRanges()

x = findOverlaps(resDf_GR, bCat_Lef1_GR)
x = unique(queryHits(x))
bCat_Lef1_peaks = as.data.frame(resDf_GR[x])
bCat_Lef1_peaks$peak = paste(bCat_Lef1_peaks$seqnames, bCat_Lef1_peaks$start, sep=":") %>% paste(bCat_Lef1_peaks$end, sep="-")
bCat_Lef1_peaks
```

Scatter plot of bCat-bound peaks
```{r}
bCat_scatter = ggplot() + theme_classic() + 
  geom_hline(yintercept=0, linetype="dashed", size=0.5) +
  geom_vline(xintercept=0, linetype="dashed", size=0.5) +
  labs(x="DMSO vs CHIR", y="DMSO vs XAV") +
  geom_point(data=resDf, aes(x=log2FoldChange.x, y=log2FoldChange.y), alpha=0.4, color="gray") +
  geom_point(data=dplyr::filter(resDf, peak %in% bCat_Lef1_peaks$peak), 
             aes(x=log2FoldChange.x, y=log2FoldChange.y), alpha=0.5, color="#9D0041") +
  lims(x=c(-4.5,3.5), y=c(-3.5,4))
bCat_scatter

options(scipen = 0)
#Association with wnt-responsive peaks
nrow(downCHIR_upXAV[downCHIR_upXAV$peak %in% bCat_Lef1_peaks$peak,]) #13 #0 #2
nrow(downCHIR_upXAV[!downCHIR_upXAV$peak %in% bCat_Lef1_peaks$peak,]) #1875 #1888 #1886
nrow(upCHIR_downXAV[upCHIR_downXAV$peak %in% bCat_Lef1_peaks$peak,]) #187 #52 #165
nrow(upCHIR_downXAV[!upCHIR_downXAV$peak %in% bCat_Lef1_peaks$peak,]) #1639 #1774 #1661

test = as.table(rbind(c(2, 165), c(1886, 1661)))
dimnames(test) = list(bCat_bound = c("YES","NO"),
                       peaks = c("downCHIR_upXAV", "upCHIR_downXAV"))
chisq = chisq.test(test)
chisq$p.value
```

```{r}
DMSO_CHIR_resDf = DMSO_CHIR_resDf %>% 
  arrange(log2FoldChange) %>% 
  mutate(rank = 1:nrow(DMSO_CHIR_resDf)) %>% 
  mutate(bCat_peak = peak %in% bCat_Lef1_peaks$peak) %>%
  mutate(anterior = peak %in% anterior_peaks$peak) %>%
  mutate(posterior = peak %in% posterior_peaks$peak)

DMSO_XAV_resDf = DMSO_XAV_resDf %>% 
  arrange(log2FoldChange) %>% 
  mutate(rank = 1:nrow(DMSO_XAV_resDf)) %>% 
  mutate(bCat_peak = peak %in% bCat_Lef1_peaks$peak) %>%
  mutate(anterior = peak %in% anterior_peaks$peak) %>%
  mutate(posterior = peak %in% posterior_peaks$peak)
```

Row heatmaps
```{r}
colors = colorRampPalette(c("#9D0041","#9D0041","#E7E8E9","#054D80","#054D80"))(100)
DMSO_CHIR_heatmap_1=ggplot(DMSO_CHIR_resDf, aes(x=rank, y=1, color=log2FoldChange)) + geom_tile() +
  scale_color_gradientn(values = c(1,0.72,0.5,0), colours=colors) +
    theme_classic() +
  theme(legend.position = "top") 
DMSO_CHIR_heatmap_1

DMSO_CHIR_heatmap_2=ggplot(filter(DMSO_CHIR_resDf, bCat_peak==TRUE), aes(x=rank, y=1, color=bCat_peak, fill=bCat_peak)) + 
  geom_tile(fill=alpha("black", 0.1), color = alpha("black", 0.2)) +
  theme_classic() +
  theme(legend.position = "top") 
DMSO_CHIR_heatmap_2


colors = colorRampPalette(c("#054D80","#054D80","#E7E8E9","#9D0041","#9D0041"))(100)
DMSO_XAV_heatmap_1=ggplot(DMSO_XAV_resDf, aes(x=rank, y=1, color=log2FoldChange)) + geom_tile() +
  scale_color_gradientn(values = c(1,0.47,0), colours=colors) +
    theme_classic() +
  theme(legend.position = "top", ) 
DMSO_XAV_heatmap_1

DMSO_XAV_heatmap_2=ggplot(filter(DMSO_XAV_resDf, bCat_peak==TRUE), aes(x=rank, y=1, color=bCat_peak, fill=bCat_peak)) + 
  geom_tile(fill=alpha("black", 0.1), color = alpha("black", 0.2)) +
  theme_classic() +
  theme(legend.position = "top") 
DMSO_XAV_heatmap_2

#association with wnt GOF
test = as.table(rbind(c(105, 919), c(35271, 25277)))
dimnames(test) <- list(bCat_bound = c("YES","NO"),
                       peaks = c("downCHIR", "upCHIR"))
chisq = chisq.test(test)
chisq$p.value

#association with wnt LOF
test = as.table(rbind(c(410, 147), c(24630, 28055)))
dimnames(test) = list(bCat_bound = c("YES","NO"),
                       peaks = c("downXAV", "upXAV"))
chisq = chisq.test(test) #HH5:X-squared = 46.808, df = 1, p-value = 0.00000000000783
chisq$p.value
```

Boxplots of anterior/posterior peaks
```{r}
norm_counts = norm_counts %>%
  mutate(peak = paste0(CHR, ":", START, "-", END)) %>%
  dplyr::select(-CHR, -START, -END) %>% remove_rownames() %>%
  column_to_rownames(var = c("peak"))

norm_counts_av = norm_counts %>% mutate(Epi_DMSO_av = rowMeans(norm_counts[,c(1:3)])) %>%
  mutate(Epi_CHIR_av = rowMeans(norm_counts[,c(4:6)])) %>% 
  mutate(Epi_XAV_av = rowMeans(norm_counts[,c(7:9)]))
norm_counts_av = norm_counts_av[,c(10:13)]

norm_counts_av_long = norm_counts_av %>% gather("condition", "counts", -peak)
norm_counts_av_long = norm_counts_av_long %>%
  mutate(condition = str_sub(norm_counts_av_long$condition, 5,-4))
norm_counts_av_long$condition = factor(norm_counts_av_long$condition, 
    levels=c("DMSO", "CHIR", "XAV"))

anterior_boxplot = ggplot(norm_counts_av_long %>% dplyr::filter(peak %in% anterior_peaks$peak),
       aes(x=condition, y=log2(counts), fill=condition)) + 
  geom_boxplot(notch = T) + theme_classic() + scale_fill_brewer(palette="Blues")
anterior_boxplot

posterior_boxplot = ggplot(norm_counts_av_long %>% dplyr::filter(peak %in% posterior_peaks$peak),
       aes(x=condition, y=log2(counts), fill=condition)) + 
  geom_boxplot(notch = T) + theme_classic() + scale_fill_brewer(palette="Reds")
posterior_boxplot
```







